src/python_be.cc

```diff
@@ -1542,11 +1593,20 @@ ModelInstanceState::ProcessRequests(
                   pb_memory->ByteSize(), pb_memory->DataPtr(), pointer,
                   CudaStream(), &cuda_used));
           cuda_copy |= cuda_used;
+        } else if (
+            pb_memory->MemoryType() == TRITONSERVER_MEMORY_GPU &&
+            pb_memory->UseCudaSharedPool()) {
+#ifdef TRITON_ENABLE_GPU
+          // Copy the data from the CUDA shared memory pool to the output buffer
+          // provided by Triton
+          pb_memory->WriteBackGPUOutput(Stub()->ShmPool(), stream_);
+          write_back = true;
+#endif  // TRITON_ENABLE_GPU
```

Comment:
q: why is the data copy happening here instead of the stub process?

Reply:
This is after the stub process copies the GPU tensors to the parent-provided buffer. We need to copy the result back to Triton provided buffer so that the output is properly passed back to Triton.

Reply:
I thought if the memory buffer returned by Triton is not from GPU memory pool, we would revert back to the previous cudaIpc policy. Is this extra copy more performant than opening/closing cudaIpc handle?

Reply:
> Is this extra copy more performant than opening/closing cudaIpc handle?

Yes, from the nsight traces, it appears that doing the extra copy has better performance.
Originally, the time for the stub to open cudaIpc handle, copy output to the buffer, and close the cudaIpc handle, is about  550 microseconds.
 _(red: cuda copy; green: open/close cuda handle)_
![original_screenshot](https://github.com/triton-inference-server/python_backend/assets/43719498/79143573-e56b-40d7-ac7e-f5062eadf4ec)

After the optimization, the time for the stub to copy output to the intermediate buffer, which is created by the parent from GPU memory pool (the bottom section in the trace), and the time for the parent to do an extra copy from that intermediate buffer to Triton-provided buffer (the upper section in the trace), is about 350 microseconds. 
_(red: cuda copy; green: cuda synchronization)_
![new_screenshot](https://github.com/triton-inference-server/python_backend/assets/43719498/b447dff5-9672-4a71-ac9b-fe06b6ccd843)